### PR TITLE
CON-178, CON-226, CON-227, CON-184

### PR DESCRIPTION
**Pull request by samuelraeburn**
CON-226: Added npm scripts to package.json
    
    - Via npm it is now possible to run the tests, code coverage and linter
    - Fixed many lint issues
    
    CON-227: Fixed issue where a promise was never resolved
    
    - When adding tests in order to increase code coverage this issue
      was discovered, and fixed. The promise returned by output.wait
      is now correctly rejected or resolved.
    
    CON-178: Added code coverage via istanbul package
    - Added tests to achieve following code coverage stats:
        Statements   : 91.33% ( 358/392 )
        Branches     : 81.41% ( 162/199 )
        Functions    : 100% ( 89/89 )
        Lines        : 91.33% ( 358/392 )
    
    With the above changes we can vastly improve the CI we have.
    Additionally, we can now test against multiple versions of Node.JS by
    installing nvm on the test machines and doing:
    nvm install 10
    nvm use 10
    npm run test-ci coverage-ci lint
    nvm install 12
    nvm use 12
    npm run test-ci coverage-ci lint
    ...

### CODE REVIEW

**Comment by samuelraeburn** [comment]
Wait for https://github.com/rticommunity/rticonnextdds-connector-js/pull/58 to be merged

File: test/nodejs/test_rticonnextdds_connector.js [hunk 1]
```diff
@@ -70,64 +70,107 @@ describe('Connector Tests', function () {
...
+    const connector = new rti.Connector(
+      'MyParticipantLibrary::ConnectorWithParticipantQos',
+      xmlProfile)
+    expect(connector).to.exist
+    expect(connector).to.be.instanceOf(rti.Connector)
   })
 
-  describe('Connector callback test', function () {
-    let connector
+  it('Is possible to set max objects per thread', function () {
```
Comment (samuelraeburn) [inline test/nodejs/test_rticonnextdds_connector.js:104]: :point_up: This is a new test

File: test/nodejs/test_rticonnextdds_connector.js [hunk 2]
```diff
@@ -70,64 +70,107 @@ describe('Connector Tests', function () {
...
+      done()
+    }, 1000)
+    connector.once('error', errorSpy)
+    // Need to cause the onDataAvailable callback to throw an error, we do
+    // this by concurrently waiting on the same connector object
+    connector.wait(500)
+    connector.once('on_data_available', () => {})
+  })
+
+  it('internal waitset is waited on repeatedly within on_data_available', function (done) {
```
Comment (samuelraeburn) [inline test/nodejs/test_rticonnextdds_connector.js:156]: :point_up: This is a new test

File: test/nodejs/test_rticonnextdds_connector.js [hunk 3]
```diff
@@ -70,64 +70,107 @@ describe('Connector Tests', function () {
...
+      done() // Pattern for async testing: next test won't execute until done gets called.
+    }, 1000) // Expectation Test will execute after 1000 milisec
+    connector.once('on_data_available', spy)
+    output = connector.getOutput('MyPublisher::MySquareWriter')
+    testMsg = '{"x":1,"y":1,"z":true,"color":"BLUE","shapesize":5}'
+    output.instance.setFromJson(JSON.parse(testMsg))
+    output.write()
+  })
+
+  it('on_data_available emits the error event on error', function (done) {
```
Comment (samuelraeburn) [inline test/nodejs/test_rticonnextdds_connector.js:141]: point_up This is a new test

File: test/nodejs/test_rticonnextdds_data_access.js [hunk 4]
```diff
@@ -101,16 +101,64 @@ describe('Data access tests with a pre-populated input', function () {
     expect(sample.get('my_long')).to.deep.equals(10).and.is.a('number')
   })
 
+  it('getNumber requires a valid index', () => {
```
Comment (samuelraeburn) [inline test/nodejs/test_rticonnextdds_data_access.js:104]: :point_down: 2 new tests here

File: test/nodejs/test_rticonnextdds_data_access.js [hunk 5]
```diff
@@ -101,16 +101,64 @@ describe('Data access tests with a pre-populated input', function () {
...
+      prepopulatedInput.samples.getNumber(0, 1)
+    }).to.throw(TypeError)
+  })
+
   it('getString on a number field should return a string', () => {
     expect(sample.getString('my_long')).to.deep.equals('10').and.is.a('string')
     expect(sample.getString('my_double')).to.deep.equals('3.3').and.is.a('string')
   })
 
+  it('getString requires a valid index', () => {
```
Comment (samuelraeburn) [inline test/nodejs/test_rticonnextdds_data_access.js:121]: :point_down: 2 new tests here

File: test/nodejs/test_rticonnextdds_data_access.js [hunk 6]
```diff
@@ -101,16 +101,64 @@ describe('Data access tests with a pre-populated input', function () {
...
+      prepopulatedInput.samples.getString(0, 1)
+    }).to.throw(TypeError)
+  })
+
   it('getBoolean should return a boolean', () => {
     expect(sample.getBoolean('my_optional_bool')).to.be.true.and.is.a('boolean')
     expect(sample.get('my_optional_bool')).to.be.true.and.is.a('boolean')
   })
 
+  it('getBoolean requires a valid index', () => {
```
Comment (samuelraeburn) [inline test/nodejs/test_rticonnextdds_data_access.js:138]: :point_down: 4 new tests here

File: test/nodejs/test_rticonnextdds_data_access.js [hunk 7]
```diff
@@ -199,6 +247,18 @@ describe('Data access tests with a pre-populated input', function () {
     }).to.throw(rti.DDSError)
   })
 
+  it('getJson requires valid index', () => {
```
Comment (samuelraeburn) [inline test/nodejs/test_rticonnextdds_data_access.js:250]: :point_down: 2 new tests

File: test/nodejs/test_rticonnextdds_data_access.js [hunk 8]
```diff
@@ -344,8 +404,14 @@ describe('Data access tests with a pre-populated input', function () {
   })
 
   it('Obtain JSON string of dictionary', () => {
-      const jsonInstance = output.instance.getJson()
-      expect(jsonInstance).to.deep.equals(testJsonObject)
+    const jsonInstance = output.instance.getJson()
+    expect(jsonInstance).to.deep.equals(testJsonObject)
+  })
+
+  it('samples.getNative requires valid index', () => {
```
Comment (samuelraeburn) [inline test/nodejs/test_rticonnextdds_data_access.js:411]: :point_down: This is a new test

File: test/nodejs/test_rticonnextdds_data_access.js [hunk 9]
```diff
@@ -864,6 +930,59 @@ describe('Tests with a testOutput and testInput', () => {
     expect(sample.getNumber('my_point_sequence#')).to.deep.equals(2)
   })
 
+  it('Can clear an entire instance on an output', async () => {
```
Comment (samuelraeburn) [inline test/nodejs/test_rticonnextdds_data_access.js:933]: :point_down: New test

File: test/nodejs/test_rticonnextdds_discovery.js [hunk 10]
```diff
@@ -405,4 +405,32 @@ describe('Discovery tests', function () {
     //  since they were all created from the same DomainParticipant as output,
     //  which will have delete_contained_entities called on it.
   })
+
+  it('waitForPublications timeout defaults to infinity', async function () {
```
Comment (samuelraeburn) [inline test/nodejs/test_rticonnextdds_discovery.js:409]: :point_down: New tests
I am not sure if we should keep this or not.
If something were to break, the tests will hang (infinite wait) I added for code coverage purpose.

File: rticonnextdds-connector.js [hunk 11]
```diff
@@ -1589,10 +1592,15 @@ class Output {
         timeout,
         (err, res) => {
           if (err) {
-            reject(err)
+            return reject(err)
```
Comment (samuelraeburn) [inline rticonnextdds-connector.js:1595]: :point_down: This was broken previously.

**Comment by samuelraeburn** [comment]
@alexcamposruiz This PR has a lot of changes in, because I fixed linting problems with indentation.
I have gone through the PR and added a comment to any non-linting changes.

File: package.json [hunk 12]
```diff
@@ -33,5 +41,17 @@
   "bugs": {
     "url": "https://github.com/rticommunity/rticonnextdds-connector-js/issues"
   },
-  "homepage": "https://github.com/rticommunity/rticonnextdds-connector-js"
+  "homepage": "https://github.com/rticommunity/rticonnextdds-connector-js",
+  "devDependencies": {
```
Comment (samuelraeburn) [inline package.json:45]: I have added the devDependencies which we have, which seemed like a best practice

File: package.json [hunk 13]
```diff
@@ -19,6 +19,14 @@
     "ffi-napi": "^3.1.0",
     "sleep": "^6.3.0"
   },
+  "scripts": {
```
Comment (samuelraeburn) [inline package.json:22]: These scripts can be run via `npm run <script>`.
My idea is that in Jenkins we can just do
`nvm use 10 && npm run coverage-ci && nvm use 12 && npm run coverage-ci` etc.

**Comment by samuelraeburn** [comment]
Also did CON-184 (pdf docs). Generated PDF attached
[rticonnectorforjavascript.pdf](https://github.com/rticommunity/rticonnextdds-connector-js/files/6538189/rticonnectorforjavascript.pdf)

File: test/nodejs/test_rticonnextdds_connector.js (hunk 1 above)
Comment (alexcamposruiz) [reply test/nodejs/test_rticonnextdds_connector.js:104]: Aren't we removing this option because it's no longer necessary in Core?

**Review by alexcamposruiz — APPROVED** [review]

**Pull request merged**